src/pykeen/nn/node_piece/anchor_search.py

```diff
@@ -222,6 +222,152 @@ def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) ->
+            from torch_sparse import spmm
+        except ImportError as err:
+            raise ImportError(f"Requires `torch_sparse` to be installed.") from err
+
+        num_entities = edge_list.max().item() + 1
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        device = resolve_device()
+        anchors = torch.tensor(anchors, dtype=torch.long, device=device)
+
+        # an array storing whether node i is reachable by anchor j
+        reachable = torch.zeros((num_entities, num_anchors), dtype=torch.bool, device=device)
+        reachable[anchors] = torch.eye(num_anchors, dtype=torch.bool, device=device)
+
+        # an array indicating whether a node is closed, i.e., has found at least $k$ anchors
+        final = torch.zeros((num_entities,), dtype=torch.bool, device=device)
+
+        # the output that track the distance to each found anchor
+        # dtype is 8 bit, so we initialize the maximum distance to 127
+        # TODO int8 takes 8x more space than just a binary matrix, decide if we want to keep it
+        pool = torch.zeros((num_entities, num_anchors), dtype=torch.int8, device=device).fill_(127)
+        # initial anchors are 0-hop away from themselves
+        pool[anchors, torch.arange(len(anchors), dtype=torch.long, device=device)] = 0
+
+        edge_list = edge_list.to(device)
+        values = torch.ones_like(edge_list[0], dtype=torch.bool, device=device)
+
+        # TODO: take all (q-1) hop neighbors before selecting from q-hop
+        old_reachable = reachable
+        for i in range(max_iter):
+            # propagate one hop
+            reachable = spmm(index=edge_list, value=values, m=num_entities, n=num_entities, matrix=reachable)
+            # convergence check
+            if (reachable == old_reachable).all():
+                logger.warning(f"Search converged after iteration {i} without all nodes being reachable.")
+                break
+            # newly reached is a mask that points to newly discovered anchors at this particular step
+            # implemented as element-wise XOR (will only give True in 0 XOR 1 or 1 XOR 0)
+            # in our case we enrich the set of found anchors, so we can only have values turning 0 to 1, eg 0 XOR 1
+            newly_reached = reachable ^ old_reachable
+            old_reachable = reachable
+            # copy pool if we have seen enough anchors and have not yet stopped
+            num_reachable = reachable.sum(axis=1)
+            enough = num_reachable >= k
+            mask = enough & ~final
+            logger.debug(
+                f"Iteration {i}: {format_relative_comparison(enough.sum(), total=num_entities)} closed nodes.",
+            )
+            # update the value in the pool by the current hop value (we start from 0, so +1 be default)
+            pool[newly_reached] = i + 1
+            # stop once we have enough
+            final |= enough
+            if final.all():
+                break
```

Comment:
```suggestion
            if final.all():
                break
        else:
            logger.warning(f"Could not close {format_relative_comparison((~final).sum().item(), total=num_entities)} } nodes after {max_iter} hops")
```

Reply:
technically, we already have this warning in the main tokenization class 
https://github.com/pykeen/pykeen/blob/e49a3551f4a54ea246a5233ce6d0aa0291f1c313/src/pykeen/nn/node_piece/tokenization.py#L146-L149

sometimes those warning/debug messages indeed are not printed

Reply:
yes, for instance, I don't see those debug messages 
https://github.com/pykeen/pykeen/blob/e49a3551f4a54ea246a5233ce6d0aa0291f1c313/src/pykeen/nn/node_piece/anchor_search.py#L183-L185

even when setting `logging.basicConfig(level=logging.DEBUG)` in my script

Reply:
could you try setting 
```python
logging.getLogger("pykeen").setLevel(level=logging.DEBUG)
```

Reply:
I can see the debug messages only when setting a combination of those two: 

```python
logging.getLogger("pykeen").setLevel(level=logging.DEBUG)
logging.basicConfig(level=logging.DEBUG)
```

Reply:
Not really, it will be a duplicating piece of code exactly as we already have in the main tokenization class 
https://github.com/pykeen/pykeen/blob/e49a3551f4a54ea246a5233ce6d0aa0291f1c313/src/pykeen/nn/node_piece/tokenization.py#L146-L149

No need to repeat it I think

Reply:
okay gutheim